R/randomForest_tidiers.R

```diff
@@ -1749,3 +1749,133 @@ tidy.rpart <- function(x, type = "importance", pretty.name = FALSE, ...) {
+
+  each_func <- function(df) {
+    tryCatch({
+      clean_df_ret <- cleanup_df_per_group(df, clean_target_col, max_nrow, clean_cols, name_map, predictor_n)
+      if (is.null(clean_df_ret)) {
+        return(NULL) # skip this group
+      }
+      df <- clean_df_ret$df
+      c_cols <- clean_df_ret$c_cols
+      name_map <- clean_df_ret$name_map
+
+      # apply smote if this is binary classification
+      unique_val <- unique(df[[clean_target_col]])
+      if (smote && length(unique_val[!is.na(unique_val)]) == 2) {
+        df <- df %>% exp_balance(clean_target_col, sample=FALSE) # since we already sampled, no further sample.
+      }
+
+      # build formula for randomForest
+      rhs <- paste0("`", c_cols, "`", collapse = " + ")
+      fml <- as.formula(paste(clean_target_col, " ~ ", rhs))
+      model_df <- model.frame(fml, data = df, na.action = randomForest::na.roughfix)
+
+      # all or max_sample_size data will be used for randomForest
+      # to grow a tree
+      if (is.null(max_sample_size)) { # default to half of max_nrow
+        max_sample_size = max_nrow/2
+      }
+      sample.fraction <- min(c(max_sample_size / max_nrow, 1))
+
+      rf <- Boruta::Boruta(
+        fml,
+        data = model_df,
+        doTrace = 0,
+        maxRuns = max_runs + 1, # It seems Boruta stops at maxRuns - 1 iterations. Add 1 to be less confusing.
+        pValue = p_value,
+        # importance = "impurity", # In calc_feature_imp, we use impurity, but Boruta's getImpRfZ function uses permutation.
+        # Following parameters are to be relayed to ranger::ranger through Boruta::Boruta, then Boruta::getImpRfZ.
+        num.trees = ntree,
+        min.node.size = nodesize,
+        sample.fraction = sample.fraction,
+        probability = (classification_type == "binary") # build probability tree for AUC only for binary classification.
+      )
+
+      # these attributes are used in tidy of randomForest
+      rf$classification_type <- classification_type
+      rf$orig_levels <- orig_levels
+      rf$terms_mapping <- names(name_map)
+      names(rf$terms_mapping) <- name_map
+      rf
+    }, error = function(e){
+      if(length(grouped_cols) > 0) {
+        # ignore the error if
+        # it is caused by subset of
+        # grouped data frame
+        # to show result of
+        # data frames that succeed
+        NULL
+      } else {
+        stop(e)
+      }
+    })
+  }
+
+  do_on_each_group(clean_df, each_func, name = "model", with_unnest = FALSE)
+}
+
+tidy.Boruta <- function(x, ...) {
+  res <- tidyr::gather(as.data.frame(x$ImpHistory), "variable","importance")
+  decisions <- data.frame(variable=names(x$finalDecision), decision=x$finalDecision)
+  res <- res %>% dplyr::left_join(decisions, by = "variable") 
+  res$variable <- x$terms_mapping[res$variable] # Map variable names back to original.
+  res <- res %>% dplyr::filter(decision %in% c("Confirmed", "Tentative", "Rejected"))
```

Comment:
Should describe why we need to filter by those 3 items. Also, seems it is used in multiple places so it would be better to define a constant and reuse it.

Reply:
Added comment. The list of levels appear in line 1869 and 1871, but in 1871, they has to be directly embedded in arguments. So let me keep it this way.